File: conformance/tests/p256-wallet/p256-wallet.runar.rb
require 'runar'

class P256Wallet < Runar::SmartContract
  prop :ecdsa_pub_key_hash, Addr
  prop :p256_pub_key_hash, ByteString

  def initialize(ecdsa_pub_key_hash, p256_pub_key_hash)
    super(ecdsa_pub_key_hash, p256_pub_key_hash)
    @ecdsa_pub_key_hash = ecdsa_pub_key_hash
    @p256_pub_key_hash = p256_pub_key_hash
  end

  runar_public p256_sig: ByteString, p256_pub_key: ByteString, sig: Sig, pub_key: PubKey
  def spend(p256_sig, p256_pub_key, sig, pub_key)
    assert hash160(pub_key) == @ecdsa_pub_key_hash
    assert check_sig(sig, pub_key)
    assert hash160(p256_pub_key) == @p256_pub_key_hash
    assert verify_ecdsa_p256(sig, p256_sig, p256_pub_key)
  end
end

File: examples/zig/p256-wallet/P256Wallet.runar.zig
const runar = @import("runar");

pub const P256Wallet = struct {
    pub const Contract = runar.SmartContract;

    ecdsaPubKeyHash: runar.Addr,
    p256PubKeyHash: runar.ByteString,

    pub fn init(ecdsaPubKeyHash: runar.Addr, p256PubKeyHash: runar.ByteString) P256Wallet {
        return .{
            .ecdsaPubKeyHash = ecdsaPubKeyHash,
            .p256PubKeyHash = p256PubKeyHash,
        };
    }

    pub fn spend(
        self: *const P256Wallet,
        p256Sig: runar.ByteString,
        p256PubKey: runar.ByteString,
        sig: runar.Sig,
        pubKey: runar.PubKey,
    ) void {
        runar.assert(runar.bytesEq(runar.hash160(pubKey), self.ecdsaPubKeyHash));
        runar.assert(runar.checkSig(sig, pubKey));
        runar.assert(runar.bytesEq(runar.hash160(p256PubKey), self.p256PubKeyHash));
        runar.assert(runar.verifyECDSA_P256(sig, p256Sig, p256PubKey));
    }
};
